File: src/DFRobot_EnvironmentalSensor.cpp
/*!
 * @file  DFRobot_EnvironmentalSensor.cpp
 * @brief DFRobot_EnvironmentalSensor 类的基础结构，基础方法的实现
 *
 * @copyright	Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @licence   The MIT License (MIT)
 * @author    [TangJie](jie.tang@dfrobot.com)
 * @version   V1.0
 * @date      2021-08-31
 * @url       https://github.com/cdjq/DFRobot_EnvironmentalSensor
 */

#include "DFRobot_EnvironmentalSensor.h"

#ifdef __AVR__
typedef uint16_t    platformBitWidth_t;
#else
typedef uint32_t    platformBitWidth_t;
#endif



DFRobot_EnvironmentalSensor::DFRobot_EnvironmentalSensor(uint8_t addr, TwoWire *pWire)
{
  _pWire = pWire;
  _addr = addr;
};

DFRobot_EnvironmentalSensor::DFRobot_EnvironmentalSensor(uint8_t addr, Stream *s):DFRobot_RTU(s)
{
  _s = s;
  _addr = addr;
}

int8_t DFRobot_EnvironmentalSensor::begin(uint16_t pid)
{
  delay(500);
  setTimeoutTimeMs(200);
  if(_addr > 0xF7){
    DBG("Invaild Device addr.");
  }
  if(_addr != 0){
    if(!detectDeviceAddress(_addr)){
      DBG("Device addr Error.");
      return -1;
    }

    if(getDevicePID() != pid){
      DBG("PID Error");
      return -1;
    }
    
    if(getDeviceVID() != DEVICE_VID){
          DBG("VID Error");
          return -1;
      }

  }
  return 0;
}

uint16_t DFRobot_EnvironmentalSensor::getDevicePID(void)
{
  uint8_t buf[2];
  if(_pWire){
    readReg(REG_PID, buf, 2);
  }else{
    readReg(REG_PID, buf, 2);
  }
  return (buf[0] << 8 | buf[1]);

}

uint16_t DFRobot_EnvironmentalSensor::getDeviceVID(void)
{
  uint8_t buf[2];
  if(_pWire){
    readReg(REG_VID, buf, 2);
  }else{
    readReg(REG_VID, buf, 2);
  }
  return (buf[0] << 8 | buf[1]);

}

float DFRobot_EnvironmentalSensor::getTemperature(uint8_t unist)
{
  uint8_t buffer[2];
  int16_t data;
  float temp;
  
  readReg(REG_TEMP,buffer,2);

  data = buffer[0] << 8 | buffer[1];
  if(data >= 4500){
		data -= 4500;
		temp = data / 100 + ((data % 100) * 0.01);
	}else{
		data -= 4500;
		temp = data / 100 + (data % 100 ) * 0.01;
	}	
  if(unist == TEMP_F)
  {
    temp = temp * 1.8 + 32 ;
  }
  return temp;
}

uint16_t DFRobot_EnvironmentalSensor::getHumidity(void)
{
  uint8_t buffer[2];
  int16_t humidity;

  readReg(REG_HUMIDITY,buffer, 2);
  humidity = buffer[0] << 8 | buffer[1];
  return humidity;
}

float DFRobot_EnvironmentalSensor::getUltravioletIntensity(void)
{
  uint16_t ultraviolet;
  uint8_t buffer[2];
  readReg(REG_ULTRAVIOLET_INTENSITY,buffer,2);
  ultraviolet = buffer[0] << 8 | buffer[1];
  return (ultraviolet / 100) + (ultraviolet % 100) * 0.01;
}

uint16_t DFRobot_EnvironmentalSensor::getLuminousIntensity(void)
{
  uint16_t luminous;
  uint8_t buffer[2];
  readReg(REG_LUMINOUS_INTENSITY, buffer, 2);
  luminous = buffer[0] << 8 | buffer[1];
  return luminous;
}

uint16_t DFRobot_EnvironmentalSensor::getAtmospherePressure(uint8_t units)
{
  uint16_t atmosphere;
  uint8_t buffer[2];
  readReg(REG_ATMOSPHERIC_PRESSURE, buffer, 2); 
  atmosphere = buffer[0] << 8 | buffer[1];
  if(units == KPA){
    atmosphere /=10;
  }
  return atmosphere;
}

uint16_t DFRobot_EnvironmentalSensor::getElevation(void)
{
  uint16_t elevation;
  uint8_t buffer[2];
  readReg(REG_ELEVATION, buffer, 2);
  elevation = buffer[0] << 8 | buffer[1];
  return elevation;
}

uint8_t DFRobot_EnvironmentalSensor::readReg(uint16_t reg, void *pBuf, uint8_t size)
{
  uint8_t* _pBuf = (uint8_t*)pBuf;
  
    if(pBuf == NULL){
      DBG("data error");
      return 0;
    }
  if(_pWire){
    uint8_t _reg = reg * 2;
    _pWire->begin();
    _pWire->beginTransmission(_addr);
    _pWire->write(_reg);
    _pWire->endTransmission();
    _pWire->requestFrom(_addr, size);
    for(uint8_t i = 0; i < size; i++)
      _pBuf[i] = _pWire->read();
      
    return size;
  }else{
    return readHoldingRegister(_addr, reg, _pBuf, size);
  }
}

bool  DFRobot_EnvironmentalSensor::detectDeviceAddress(uint8_t addr)
{
   if(_pWire){
   uint8_t buf[2];
   readReg(REG_DEVICE_ADDR, buf, 2);
   if(addr == ((buf[0] << 8 | buf[1]) & 0xFF))
    return true;
    
  }else{
    uint16_t ret = readHoldingRegister(addr, REG_DEVICE_ADDR);
    if((ret & 0xFF) == addr)
      return true;
  }
  return false;
}

File: python/raspberry/dfrobot_environmental_sensor.py
# -*- coding: utf-8 -*
'''!
  @file        dfrobot_environmental_sensor.py
  @brief       DFRobot_EnvironmentalSensor是一个气象传感器库
  @copyright   Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
  @licence     The MIT License (MIT)
  @author      TangJie(jie.tang@dfrobot.com)
  @version     V1.0
  @date        2021-08-31
  @url         https://github.com/DFRobot/DFRobot_EnvironmentalSensor
'''

import serial
import time
import smbus
import os
import math
import RPi.GPIO as GPIO
import math

import modbus_tk
import modbus_tk.defines as cst
from modbus_tk import modbus_rtu

I2C_MODE                  = 0x01
UART_MODE                 = 0x02
DEV_ADDRESS               = 0x22

DEVICE_VID                = 0x3343
DEVICE_ADDRESS            = 0x22

##澶ф皵鍘嬪己鐧惧笗鍗曚綅
HPA                       = 0x01
##澶ф皵鍘嬪己鍗冨笗鍗曚綅
KPA                       = 0X02
##娓╁害鍗曚綅鎽勬皬搴?
TEMP_C                    = 0X03
##娓╁害鍗曚綅鍗庢皬搴?
TEMP_F                    = 0X04

class DFRobot_Environmental_Sensor():
  '''!
    @brief 定义DFRobot_Environmental_Sensor基类
    @details 用于驱动气象传感器
  '''
  def __init__(self ,bus ,baud):
    if bus != 0:
      self.i2cbus = smbus.SMBus(bus)
      self._uart_i2c = I2C_MODE
      
    else:
      self.master = modbus_rtu.RtuMaster(serial.Serial(port="/dev/ttyAMA0",baud=9600, bytesize=8, parity='N', stopbits=1))
      self.master.set_timeout(1.0)
      self._uart_i2c = UART_MODE
  
  def _detect_device_address(self):
    '''!
      @brief 鑾峰彇浼犳劅鍣ㄥ湴鍧€
      @return  杩斿洖浼犳劅鍣ㄥ湴鍧€
    '''
    rbuf = self._read_reg(0x04,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    return data

  
  def begin(self):
    '''!
    @brief 鍒濆鍖朣EN050X浼犳劅鍣?
    @brief 鍒濆鍖朣EN050X浼犳劅鍣?
    @return 杩斿洖鍊煎垵濮嬪寲鐘舵€?
    @retval 0  鎴愬姛
    @retval -1 澶辫触
    '''
    if self._detect_device_address() != DEV_ADDRESS:
      return False
    return True

  
  def get_temperature(self,unist):
    '''!
    @brief 鑾峰彇SEN050X娓╁害鏁版嵁
    @param units 娓╁害鏁版嵁鍗曚綅閫夋嫨
    @n     TEMP_C 鎽勬皬搴?
    @n     TEMP_F 鍗庢皬搴?
    @return 杩斿洖鑾峰彇鐨勬俯搴︽暟鎹?
    '''
    rbuf = self._read_reg(0x14, 2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    temp = (-45) +((data * 175.00) / 1024.00 / 64.00)
    if(unist == TEMP_F):
      temp = temp * 1.8 + 32 
    return round(temp,2)
    
  
  def get_humidity(self):
    '''!
    @brief 鑾峰彇SEN050X婀垮害鏁版嵁 
    @return 杩斿洖鑾峰彇鐨勬箍搴︽暟鎹?
    '''
    rbuf = self._read_reg(0x16, 2)
    if self._uart_i2c == I2C_MODE:
      humidity = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      humidity = rbuf[0]
    humidity = (humidity / 1024) * 100 / 64
    return humidity
  
 
  def get_ultraviolet_intensity(self):
    '''!
    @brief 鑾峰彇SEN050X绱绾垮己搴︽寚鏁版暟鎹?
    @return 杩斿洖鑾峰彇鐨勭传澶栫嚎寮哄害鎸囨暟鏁版嵁
    '''
    rbuf = self._read_reg(0x10, 2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    outputVoltage = 3.0 * data/1024
    ultraviolet = (outputVoltage - 0.99) * (15.0 - 0.0) / (2.9 - 0.99) + 0.0 
    return round(ultraviolet,2)
      
  
  def get_luminousintensity(self):
    '''!
    @brief 鑾峰彇SEN050X鍏夌嚎寮哄害鏁版嵁 
    @return 杩斿洖鑾峰彇鐨勫厜绾垮己搴︽暟鎹?
    '''
    rbuf = self._read_reg(0x12 ,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    luminous = data * (1.0023 + data * (8.1488e-5 + data * (-9.3924e-9 + data * 6.0135e-13)))
    return round(luminous,2)

  
  def get_atmosphere_pressure(self, units):
    '''!
    @brief 鑾峰彇SEN050X澶ф皵鍘嬪己鏁版嵁 
    @param units 澶ф皵鍘嬪己鏁版嵁鍗曚綅閫夋嫨
    @n            HPA 鐧惧笗
    @n            KPA 鍗冨笗
    @return 杩斿洖鑾峰彇鐨勫ぇ姘斿帇寮烘暟鎹?
    '''
    rbuf = self._read_reg(0x18, 2)
    if self._uart_i2c == I2C_MODE:
      atmosphere = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      atmosphere = rbuf[0]
    if units == KPA:
      atmosphere /= 10
    return atmosphere

  
  def get_elevation(self):
    '''!
    @brief 鑾峰彇SEN050X娴锋嫈鏁版嵁 
    @return 杩斿洖鑾峰彇鐨勬捣鎷旀暟鎹?
    '''
    rbuf = self._read_reg(0x18, 2)
    if self._uart_i2c == I2C_MODE:
      elevation = rbuf[0] << 8 | rbuf[1]
    elif self._uart_i2c == UART_MODE:
      elevation = rbuf[0]
    elevation = 44330 * (1.0 - pow(elevation / 1015.0, 0.1903));
    return round(elevation,2)

  
        

class DFRobot_Environmental_Sensor_I2C(DFRobot_Environmental_Sensor):
  '''!
  @brief An example of an i2c interface module
  '''
  def __init__(self ,bus ,addr):
    self._addr = addr
    DFRobot_Environmental_Sensor.__init__(self,bus,0)   
    
  
  def _read_reg(self, reg_addr ,length):
    '''!
    @brief read the data from the register
    @param reg register address
    @param value read data
    '''
    try:
      rslt = self.i2cbus.read_i2c_block_data(self._addr ,reg_addr , length)
    except:
      rslt = -1
    return rslt    


class DFRobot_Environmental_Sensor_UART(DFRobot_Environmental_Sensor):
  '''!
  @brief An example of an UART interface module
  '''
  def __init__(self ,baud, addr):
    self._baud = baud
    self._addr = addr
    try:
      DFRobot_Environmental_Sensor.__init__(self,0,self._baud)
    except:
      print ("plese get root!")
   
  
  def _read_reg(self, reg_addr ,length):
    '''!
    @brief 浠庝紶鎰熷櫒璇诲嚭鏁版嵁
    '''
    return list(self.master.execute(self._addr, cst.READ_INPUT_REGISTERS, reg_addr/2, length/2))
    
    